whylog/constraints/__init__.py

```diff
@@ -2,42 +2,177 @@
+
+    @classmethod
+    def verify(cls, param_dict, group_contents):
+        """
+        Verifies constraint for given params and groups contents.
+
+        For LogReader and Teacher verification.
+        """
+        raise NotImplementedError("Subclass should implement this")
 
 
 class TimeConstraint(AbstractConstraint):
-    def __init__(self, line_earlier, line_later, min_delta=None, max_delta=None):
-        pass
+    """
+    Time delta between two dates must be greater than 'max_delta' and lower than 'min_delta'
+    """
+
+    TYPE = ConstraintType.TIME_DELTA
+
+    MIN_GROUPS_COUNT = 0
+    MAX_GROUPS_COUNT = 0
+
+    GROUP_EARLIER = 'group_earlier'
+    GROUP_LATER = 'group_later'
+    MIN_DELTA = 'min_delta'
+    MAX_DELTA = 'max_delta'
+
+    def __init__(self, param_dict, groups):
+        """
+        I.e:
+        TimeConstraint(
+            {'group_earlier': (3, 1), 'group_later': (4, 2), 'min_delta': 12, 'max_delta' 30},
+            list()
+        )
+        """
+        super(TimeConstraint, self).__init__(param_dict, groups)
+
+    def convert_to_user_constraint_intent(self):
+        return super(TimeConstraint, self).convert_to_user_constraint_intent()
 
+    @classmethod
+    def get_param_names(cls):
+        return [cls.GROUP_EARLIER, cls.GROUP_LATER, cls.MIN_DELTA, cls.MAX_DELTA]
 
-class IdenticalIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+    @classmethod
+    def get_groups_count(cls):
+        return super(TimeConstraint, cls).get_groups_count()
 
+    @classmethod
+    def verify(cls, param_dict, group_contents):
+        #TODO
+        raise NotImplementedError
 
-class AnyValueIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
 
+class IdenticalConstraint(AbstractConstraint):
+    """
+    Contents of groups must be identical.
+    """
 
-class DifferentValueIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+    TYPE = ConstraintType.IDENTICAL
 
+    MIN_GROUPS_COUNT = 2
+    MAX_GROUPS_COUNT = None
 
-class ConstIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+    def __init__(self, param_dict, groups):
+        """
+        I.e:
+        IdenticalConstraint(
+            dict(),
+            [(1, 2), (2, 4)]
+        )
+        """
+        super(IdenticalConstraint, self).__init__(param_dict, groups)
+
+    def convert_to_user_constraint_intent(self):
+        return super(IdenticalConstraint, self).convert_to_user_constraint_intent()
 
+    @classmethod
+    def get_param_names(cls):
+        return []
 
-class ValueDeltaIntervals(AbstractConstraint):
-    def __init__(self, interval_lower, interval_greater, min_delta=None, max_delta=None):
+    @classmethod
+    def get_groups_count(cls):
+        return super(IdenticalConstraint, cls).get_groups_count()
+
+    @classmethod
+    def verify(cls, param_dict, group_contents):
```

Comment:
I had another idea about this function:
`def verify(cls, effect, clues, constraint)`
and `True`/`False` as a returning value. But maybe your design is better, because - if I understand correctly - you want to extract information from effect and clues and encapsulate it in `group_contents`. Anyway, I want to discuss this, because I'm afraid that we can lose some vital information by doing that.